### PR TITLE
[SPARK-17946][PYSPARK] Python crossJoin API similar to Scala

### PR DESCRIPTION
**Pull request by srinathshankar**
## What changes were proposed in this pull request?

Add a crossJoin function to the DataFrame API similar to that in Scala. Joins with no condition (cartesian products) must be specified with the crossJoin API
## How was this patch tested?

Added python tests to ensure that an AnalysisException if a cartesian product is specified without crossJoin(), and that cartesian products can execute if specified via crossJoin()

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark before opening a pull request.

### CODE REVIEW

**Comment by srinathshankar** [comment]
cc @sameeragarwal @hvanhovell @davies @JoshRosen

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -627,6 +627,25 @@ def alias(self, alias):
         return DataFrame(getattr(self._jdf, "as")(alias), self.sql_ctx)
 
     @ignore_unicode_prefix
+    @since(2.0)
```
Comment (sameeragarwal) [inline python/pyspark/sql/dataframe.py]: shouldn't this be 2.1?

**Review by sameeragarwal — APPROVED** [review]
LGTM

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -627,6 +627,25 @@ def alias(self, alias):
         return DataFrame(getattr(self._jdf, "as")(alias), self.sql_ctx)
 
     @ignore_unicode_prefix
+    @since(2.1)
+    def crossJoin(self, other):
+        """Returns the cartesian product with another :class:`DataFrame`
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: nit: add a period.

Comment (srinathshankar) [inline python/pyspark/sql/dataframe.py]: done.

**Comment by rxin** [comment]
LGTM pending Jenkins.

**Comment by rxin** [comment]
Merging in master.

**Comment by rxin** [comment]
cc @felixcheung do we need some change for R?

**Comment by felixcheung** [comment]
@rxin In R, CrossJoin is the default when expr is empty (https://github.com/apache/spark/blob/master/R/pkg/R/DataFrame.R#L2304)
I reviewed the code and documentation I think it is sufficient

**Comment by rxin** [comment]
The issue is that we want to prevent users from shooting themselves in the foot, i.e. we want to avoid accidental cross joins. The idea is unless the user explicitly says crossJoin, we should disallow crossjoins.

**Comment by felixcheung** [comment]
That's a great point. Currently R is the same as Python in that when `joinExpr` is `NULL` (R) or `on` is `None` (Python), CrossJoin is assumed. (Python [here](https://github.com/apache/spark/pull/15493/files#diff-6fc344560230bf0ef711bb9b5573f1faR650))

Problem is by default `joinExpr = NULL` (R) and `on = None` (Python) - one approach is to change both defaults, so that when omitted, they don't default to CrossJoin.
Another approach is to additional ask for `joinType` or `how` of a new `cross_join` - but this could be a bigger change.

**Comment by rxin** [comment]
Why not just introduce a crossJoin function in R, similar to Python/Scala/Java?

We don't want to change the default join type, because it is still valid to run an inner join by specifying a predicate later using the filter operator.

**Comment by rxin** [comment]
cc @felixcheung 

We still need this. I'm going to create an upstream ticket. Can one of you take it?

**Comment by srinathshankar** [comment]
Created https://issues.apache.org/jira/browse/SPARK-18013
@felixcheung @rxin

**Comment by felixcheung** [comment]
I will take that and add my note.

**Comment by rxin** [comment]
Thanks!

**Pull request closed without merging**